src/coreclr/jit/importer.cpp

```diff
@@ -4782,6 +4788,390 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
 }
 
+GenTree* Compiler::impSRCSUnsafeIntrinsic(NamedIntrinsic        intrinsic,
+                                          CORINFO_CLASS_HANDLE  clsHnd,
+                                          CORINFO_METHOD_HANDLE method,
+                                          CORINFO_SIG_INFO*     sig)
+{
+    assert(sig->sigInst.classInstCount == 0);
+
+    switch (intrinsic)
+    {
+        case NI_SRCS_UNSAFE_Add:
+        {
+            assert(sig->sigInst.methInstCount == 1);
+
+            // ldarg.0
+            // ldarg.1
+            // sizeof !!T
+            // conv.i
+            // mul
+            // add
+            // ret
+
+            GenTree* op2 = impPopStack().val;
+            GenTree* op1 = impPopStack().val;
+            impBashVarAddrsToI(op1, op2);
+
+            GenTree* size = gtNewIconNode(info.compCompHnd->getClassSize(sig->sigInst.methInst[0]), TYP_INT);
+            impBashVarAddrsToI(size);
+            size = gtNewCastNode(genActualType(TYP_I_IMPL), size, /* uns */ false, TYP_I_IMPL);
+
+            var_types type = impGetByRefResultType(GT_MUL, /* uns */ false, &op2, &size);
+            GenTree*  tmp  = new (this, GT_CALL) GenTreeOp(GT_MUL, type, op2, size DEBUGARG(/* largeNode */ true));
+
+            type = impGetByRefResultType(GT_ADD, /* uns */ false, &tmp, &op1);
+            return gtNewOperNode(GT_ADD, type, tmp, op1);
+        }
+
+        case NI_SRCS_UNSAFE_AddByteOffset:
+        {
+            assert(sig->sigInst.methInstCount == 1);
+
+            // ldarg.0
+            // ldarg.1
+            // add
+            // ret
+
+            GenTree* op2 = impPopStack().val;
+            GenTree* op1 = impPopStack().val;
+            impBashVarAddrsToI(op1, op2);
+
+            var_types type = impGetByRefResultType(GT_ADD, /* uns */ false, &op1, &op2);
+            return gtNewOperNode(GT_ADD, type, op1, op2);
+        }
+
+        case NI_SRCS_UNSAFE_AreSame:
+        {
+            assert(sig->sigInst.methInstCount == 1);
+
+            // ldarg.0
+            // ldarg.1
+            // ceq
+            // ret
+
+            GenTree* op2 = impPopStack().val;
+            GenTree* op1 = impPopStack().val;
+
+            GenTree* tmp = gtNewOperNode(GT_EQ, TYP_INT, op1, op2);
+            return gtFoldExpr(tmp);
+        }
+
+        case NI_SRCS_UNSAFE_As:
+        {
+            assert((sig->sigInst.methInstCount == 1) || (sig->sigInst.methInstCount == 2));
+
+            // ldarg.0
+            // ret
+
+            return impPopStack().val;
+        }
+
+        case NI_SRCS_UNSAFE_AsPointer:
+        {
+            assert(sig->sigInst.methInstCount == 1);
+
+            // ldarg.0
+            // conv.u
+            // ret
+
+#if (REGSIZE_BYTES == 8)
+            bool uns = true;
+#else
+            bool uns = false;
+#endif
+
+            GenTree* op1 = impPopStack().val;
+            impBashVarAddrsToI(op1);
+
+            return gtNewCastNode(genActualType(TYP_U_IMPL), op1, uns, TYP_U_IMPL);
```

Comment:
```suggestion
            return gtNewCastNode(TYP_I_IMPL, op1, uns, TYP_U_IMPL);
```

Reply:
Same below

Reply:
Fixed the various cases to just use `TYP_I_IMPL` directly